File: ios/RNTurboImagePicker/ImageProcessor.swift
import Foundation
import UIKit

public class ImageProcessor {
    public static let shared = ImageProcessor()
    
    /// Checks license and applies evaluation watermark if license is invalid.
    /// This method is intended to be called from the RN bridge layer.
    public func applyWatermarkIfNeeded(_ image: UIImage) -> UIImage {
        if !LicenseManager.shared.isValidLicense() {
            return applyEvaluationWatermark(to: image)
        }
        return image
    }
    
    /// Applies a repeating diagonal "Turbo Image Picker" watermark pattern.
    public func applyEvaluationWatermark(to image: UIImage) -> UIImage {
        let text = "Turbo Image Picker"
        let fontSize = max(20.0, min(image.size.width, image.size.height) / 15.0)
        
        let shadow = NSShadow()
        shadow.shadowColor = UIColor.black.withAlphaComponent(0.5)
        shadow.shadowOffset = CGSize(width: 1, height: 1)
        shadow.shadowBlurRadius = 3
        
        let attributes: [NSAttributedString.Key: Any] = [
            .font: UIFont.boldSystemFont(ofSize: fontSize),
            .foregroundColor: UIColor.white.withAlphaComponent(0.25),
            .shadow: shadow
        ]
        
        let attributedText = NSAttributedString(string: text, attributes: attributes)
        let textSize = attributedText.size()
        
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1.0 // Prevent massive memory inflation on Retina devices
        let renderer = UIGraphicsImageRenderer(size: image.size, format: format)
        return renderer.image { context in
            image.draw(at: .zero)
            
            let cgContext = context.cgContext
            
            let diagonal = hypot(image.size.width, image.size.height)
            
            cgContext.saveGState()
            cgContext.translateBy(x: image.size.width / 2, y: image.size.height / 2)
            cgContext.rotate(by: -35.0 * .pi / 180.0)
            cgContext.translateBy(x: -diagonal / 2, y: -diagonal / 2)
            
            let stepX = textSize.width * 1.5
            let stepY = textSize.height * 3.5
            
            var y: CGFloat = 0
            var row = 0
            while y < diagonal + stepY {
                var x: CGFloat = 0
                let offset = (row % 2 == 0) ? 0 : stepX / 2
                while x < diagonal + stepX {
                    attributedText.draw(at: CGPoint(x: x + offset, y: y))
                    x += stepX
                }
                y += stepY
                row += 1
            }
            
            cgContext.restoreGState()
        }
    }
}

File: android/rnturboimagepicker/src/main/java/com/rnturboimagepicker/ImageProcessor.kt
package com.rnturboimagepicker

import android.content.Context
import android.graphics.Bitmap

object ImageProcessor {

    /**
     * Checks license and applies evaluation watermark if license is invalid.
     * This method is intended to be called from the RN bridge layer.
     */
    @JvmStatic
    fun applyWatermarkIfNeeded(context: Context, bitmap: Bitmap): Bitmap {
        return if (!LicenseManager.isValidLicense(context)) {
            applyEvaluationWatermark(bitmap)
        } else {
            bitmap
        }
    }

    /**
     * Applies a repeating diagonal "Turbo Image Picker" watermark pattern.
     */
    @JvmStatic
    fun applyEvaluationWatermark(bitmap: Bitmap): Bitmap {
        val mutableBitmap = bitmap.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = android.graphics.Canvas(mutableBitmap)

        val text = "Turbo Image Picker"
        val textPaint = android.text.TextPaint(android.graphics.Paint.ANTI_ALIAS_FLAG).apply {
            color = android.graphics.Color.argb(80, 255, 255, 255) // ~31% white
            textSize = Math.max(14f, Math.min(bitmap.width, bitmap.height) / 20f)
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.DEFAULT, android.graphics.Typeface.BOLD)
            setShadowLayer(4f, 2f, 2f, android.graphics.Color.argb(40, 0, 0, 0)) // light shadow
        }

        val textWidth = textPaint.measureText(text)
        val textHeight = textPaint.descent() - textPaint.ascent()

        val diagonal = Math.hypot(bitmap.width.toDouble(), bitmap.height.toDouble()).toFloat()

        canvas.save()
        canvas.translate(bitmap.width / 2f, bitmap.height / 2f)
        canvas.rotate(-35f)
        canvas.translate(-diagonal / 2f, -diagonal / 2f)

        val stepX = textWidth * 1.5f
        val stepY = textHeight * 3.5f

        var y = 0f
        var row = 0
        while (y < diagonal + stepY) {
            var x = 0f
            val offset = if (row % 2 == 0) 0f else stepX / 2f
            while (x < diagonal + stepX) {
                canvas.drawText(text, x + offset, y, textPaint)
                x += stepX
            }
            y += stepY
            row++
        }

        canvas.restore()
        return mutableBitmap
    }
}
